input-profile.ts

```diff
@@ -0,0 +1,369 @@
+     */
+    @property.string()
+    customProfileFolder!: string;
+
+    /**
+     * If true, the input profile will be mapped to the default controller, and no dynamic 3D model of controller will be loaded.
+     */
+    @property.bool(false)
+    mapToDefaultController!: boolean;
+
+    /**
+     * The object which has HandTracking component added to it.
+     */
+    @property.object()
+    trackedHand!: Object3D;
+
+    /**
+     * If true, adds a VR mode switch component to the loaded controller model.
+     */
+    @property.bool(true)
+    addVrModeSwitch!: boolean;
+    start() {
+        this._controllerModel = null;
+        this.toFilter = new Set(['vr-mode-active-mode-switch']);
+        this._defaultControllerComponents = this._getComponents(this.defaultController);
+        this._handedness = hands[this.handednessIndex];
+
+        this.engine.onXRSessionStart.add(() => {
+            this.engine.xr?.session.addEventListener(
+                'inputsourceschange',
+                this._onInputSourcesChange.bind(this)
+            );
+        });
+    }
+
+    onDeactivate() {
+        this.engine.xr?.session?.removeEventListener(
+            'inputsourceschange',
+            this._onInputSourcesChange.bind(this)
+        );
+    }
+
+    /**
+     * Sets newly loaded controllers for the HandTracking component to proper switching.
+     * @param {Object3D} controllerObject - The controller object.
+     * @hidden
+     */
+
+    private _setHandTrackingControllers(controllerObject: Object3D) {
+        const handtrackingComponent = this.trackedHand.getComponent(HandTracking);
+        if (!handtrackingComponent) return;
+        /** @todo: Remove any when HandTracking is typed. */
+        (handtrackingComponent as any).controllerToDeactivate = controllerObject;
+    }
+
+    /**
+     * Retrieves all components from the specified object and its children.
+     * @param {Object3D} obj - The object to retrieve components from.
+     * @return {Component[]} An array of components.
+     * @hidden
+     */
+    private _getComponents(obj: Object3D) {
+        if (obj == null) return;
+        const components: Component[] = [];
+        const stack = [obj];
+        while (stack.length > 0) {
+            const currentObj = stack.pop();
+            const comps = currentObj
+                .getComponents()
+                .filter((c: Component) => !this.toFilter.has(c.type));
+            components.push(...comps);
+            const children = currentObj.children || [];
+            // Push children onto the stack in reverse order to maintain the correct order
+            for (let i = children.length - 1; i >= 0; --i) {
+                stack.push(children[i]);
+            }
+        }
+        return components;
+    }
+
+    /**
+     * Activates or deactivates components based on the specified boolean value.
+     * @param {boolean} active - If true, components are set to active; otherwise, they are set to inactive.
```

Comment:
Again, remove `{boolean}` here and other param types elsewhere. Because it's TS, TSDoc knows.